File: 261-Graph-Valid-Tree/solution.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @return {boolean}
 */
var validTree = function(n, edges) {
    //DFS using hashmap can not detect circle like
    //[0,1], [1,2],[0,2]
    //to solve this problem, we use sth similar to UF
    
    var roots = []; //roots[i] means the root of node i
    for(let i = 0; i < n; i++)
        roots.push(i);
    for(let i = 0; i < edges.length; i++){
        var start = edges[i][0];
        var end = edges[i][1];
        
        //find the final root nodes of current start and current end
        while(start != roots[start])
            start = roots[start];
        while(end !=roots[end])
            end = roots[end];
            
        //if current start and current end ends with the same node,
        //it means there is a circle
        //so return false
        if(start === end)
            return false;
        roots[start] = end;
    }
    
    //return true;
    //we can not return true directly, for case 
    //4  [[0,1],[2,3]], there is no circle, but it can not make a graph tree
    //because number of edges may not enough make a true
    //ie there shuold be at least n-1 edges if there are n nodes
    return edges.length >= n-1; // or we can add a judge at the beginning
};

File: 261-Graph-Valid-Tree/solution.java
public class Solution {
    public boolean validTree(int n, int[][] edges) {
        int[] root = new int[n]; //root[i] is i's parent
        // Arrays.fill(root, -1);
        for(int i = 0; i < n; i++)
            root[i] = i;
        
        for(int[] pair:edges){
            int n1 = pair[0];
            int n2 = pair[1];
            
            int r1 = find(root, n1);
            int r2 = find(root, n2);
            
            if(r1 == r2)    return false;
            else{
                if(r1 < r2) root[r1] = r2;
                else    root[r2] = r1;
            }
        }
        //can not return true directly
        //for a valid graph tree, there are two conditions
        //1.no circle
        //2.all nodes are connected, not isolated
        return n-1 == edges.length;
    }
    
    public int find(int[] root, int target){
        while(target != root[target]){
            int father = root[root[target]];
            root[target] = father;
            target = root[target];
        }
        return target;
    }
}